otherlibs/runtime_events/runtime_events_consumer.c

```diff
@@ -88,57 +88,47 @@ struct caml_runtime_events_cursor {
 
 /* C-API for reading from an runtime_events */
 
-runtime_events_error
-caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
-                             struct caml_runtime_events_cursor **cursor_res) {
-  int ret;
-
-#ifndef _WIN32
-  int ring_fd;
-  struct stat tmp_stat;
-#endif
-
-  struct caml_runtime_events_cursor *cursor =
-      caml_stat_alloc_noexc(sizeof(struct caml_runtime_events_cursor));
+/** Return a new string with the path of the ring file */
+static runtime_events_error format_runtime_events_loc(
+  const char_os *input_path, int input_pid,
+  char_os **out_runtime_events_loc
+) {
   char_os *runtime_events_loc;
 
-  if (cursor == NULL) {
-    return E_ALLOC_FAIL;
-  }
-
-  runtime_events_loc = caml_stat_alloc_noexc(RING_FILE_NAME_MAX_LEN);
-
-  if (runtime_events_loc == NULL) {
-    caml_stat_free(cursor);
-    return E_ALLOC_FAIL;
-  }
-
-  /* If pid < 0 then we create a cursor for the current process */
-  if (pid < 0) {
+  if (input_pid < 0) {
+    /* Attaching to this process' ring, if it exists */
     runtime_events_loc = caml_runtime_events_current_location();
+    if (runtime_events_loc == NULL) return E_NO_CURRENT_RING;
 
-    if( runtime_events_loc == NULL ) {
-      caml_stat_free(cursor);
-      return E_NO_CURRENT_RING;
-    }
+    /* return a new string every time */
+    runtime_events_loc = caml_stat_strdup_os(runtime_events_loc);
   } else {
-  /* In this case we are reading the ring for a different process */
-    if (runtime_events_path) {
-      ret = snprintf_os(runtime_events_loc, RING_FILE_NAME_MAX_LEN,
-                        T("%s/%d.events"), runtime_events_path, pid);
+    /* Attaching to a process by directory and PID */
+    int err;
+    runtime_events_loc = caml_stat_alloc_noexc(RING_FILE_NAME_MAX_LEN);
+    if (runtime_events_loc == NULL) return E_ALLOC_FAIL;
+
+    if (input_path) {
+      err = snprintf_os(runtime_events_loc, RING_FILE_NAME_MAX_LEN,
+                        T("%s/%d.events"), input_path, input_pid);
     } else {
-      ret =
-          snprintf_os(runtime_events_loc, RING_FILE_NAME_MAX_LEN,
-                      T("%d.events"), pid);
+      err = snprintf_os(runtime_events_loc, RING_FILE_NAME_MAX_LEN,
+                        T("%d.events"), input_pid);
     }
-
-    if (ret < 0) {
-      caml_stat_free(cursor);
+    if (err < 0) {
       caml_stat_free(runtime_events_loc);
       return E_PATH_FAILURE;
     }
   }
 
+  *out_runtime_events_loc = runtime_events_loc;
+  return E_SUCCESS;
+}
+
+static runtime_events_error
+cursor_map_ring_file(struct caml_runtime_events_cursor *cursor,
+                     char_os *runtime_events_loc) {
+  int ret = 0;
 #ifdef _WIN32
   cursor->ring_file_handle = CreateFile(
```

Comment:
Strong preference for not setting any fields in `cursor` until we know we're succeeding. Use local variables and then copy them in right before `return E_SUCCESS`.